File: src/jp/lionas/alexa/fundra/mufg/FundUtil.java
package jp.lionas.alexa.fundra.mufg;

import com.amazon.ask.dispatcher.request.handler.HandlerInput;
import com.amazon.ask.model.Response;
import com.google.gson.Gson;
import jp.lionas.alexa.fundra.Const;
import jp.lionas.alexa.fundra.mufg.model.CodeItem;
import jp.lionas.alexa.fundra.mufg.model.FundItem;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public class FundUtil {

    private static String getLatestFundStatus(CodeItem item) {

        String result;

        try {
            List<FundItem> funds = MufgFund.getLatestFund(item.getFundCode());
            if(funds == null || funds.isEmpty()) {
                return Const.ERROR;
            }

            FundItem fundItem = funds.get(0);
            result = String.format(Const.STATUS_MESSAGE,
                    fundItem.getFundName(),
                    fundItem.getNav(),
                    fundItem.getCmpPrevDay() >= 0 ? Const.PLUS : Const.MINUS,
                    Math.abs(fundItem.getCmpPrevDay()));

        } catch (Exception e) {
            return Const.ERROR;
        }

        return result;

    }

    public static Optional<Response> getLastFund(HandlerInput handlerInput,
                                                 Map<String, Object> sessionAttributes,
                                                 CodeItem lastItem) {

        // 1件だけマッチした場合
        sessionAttributes.put(Const.STATE_KEY, Const.STATE_END);
        String speechText = FundUtil.getLatestFundStatus(lastItem);
        return handlerInput.getResponseBuilder()
                .withSpeech(speechText)
                .withSimpleCard(Const.SKILL_NAME, speechText)
                .withShouldEndSession(true)
                .build();

    }

    public static Optional<Response> getRepeatedFunds(HandlerInput handlerInput,
                                                      Map<String, Object> sessionAttributes,
                                                      List<CodeItem> found) {

        String speechText = "";
        int targetIndex = (int) sessionAttributes.get(Const.INDEX_KEY);
        if(targetIndex == 0) {
            speechText += String.format(Const.FOUND_MULTIPLE, found.size());
        }
        sessionAttributes.put(Const.INDEX_KEY, targetIndex+1);

        if(targetIndex < found.size()-1) {
            // 最後の要素でない場合
            sessionAttributes.put(Const.STATE_KEY, Const.STATE_REPEAT);
            speechText += FundUtil.getLatestFundStatus(found.get(targetIndex));
            speechText += Const.ASK_NEXT;
            return handlerInput.getResponseBuilder()
                    .withSpeech(speechText)
                    .withSimpleCard(Const.SKILL_NAME, speechText)
                    .withReprompt(Const.ASK_NEXT)
                    .withShouldEndSession(false)
                    .build();
        } else {
            // 最後の要素の場合
            return getLastFund(handlerInput, sessionAttributes, found.get(targetIndex));
        }

    }

}

File: src/jp/lionas/alexa/fundra/mufg/FundUtil.kt
package jp.lionas.alexa.fundra.mufg

import com.amazon.ask.dispatcher.request.handler.HandlerInput
import com.amazon.ask.model.Response
import jp.lionas.alexa.fundra.def.Message
import jp.lionas.alexa.fundra.def.State
import jp.lionas.alexa.fundra.mufg.model.CodeItem
import java.util.Optional

object FundUtil {

    private fun getLatestFundStatus(item: CodeItem): String? {

        try {
            val fundCode: String? = item.fundCode

            fundCode?.let{
                val funds = MufgFund.getLatestFund(it)
                if (funds == null || funds.isEmpty()) {
                    return Message.ERROR
                }

                val fundItem = funds[0]
                return String.format(Message.STATUS_MESSAGE,
                        fundItem.fundName,
                        fundItem.nav,
                        if (fundItem.cmpPrevDay >= 0) Message.PLUS else Message.MINUS,
                        Math.abs(fundItem.cmpPrevDay))
            }

        } catch (e: Exception) {
            return Message.ERROR
        }

        return ""

    }

    fun getLastFund(handlerInput: HandlerInput,
                    sessionAttributes: MutableMap<String, Any>,
                    lastItem: CodeItem): Optional<Response> {

        // 1件だけマッチした場合
        sessionAttributes[State.STATE_KEY] = State.STATE_END
        val speechText = FundUtil.getLatestFundStatus(lastItem)
        return handlerInput.responseBuilder
                .withSpeech(speechText)
                .withSimpleCard(Message.SKILL_NAME, speechText)
                .withShouldEndSession(true)
                .build()

    }

    fun getRepeatedFunds(handlerInput: HandlerInput,
                         sessionAttributes: MutableMap<String, Any>,
                         found: List<CodeItem>): Optional<Response> {

        var speechText = ""
        val targetIndex = sessionAttributes[State.INDEX_KEY] as Int
        if (targetIndex == 0) {
            speechText += String.format(Message.FOUND_MULTIPLE, found.size)
        }
        sessionAttributes[State.INDEX_KEY] = targetIndex + 1

        return if (targetIndex < found.size - 1) {
            // 最後の要素でない場合
            sessionAttributes[State.STATE_KEY] = State.STATE_REPEAT
            speechText += FundUtil.getLatestFundStatus(found[targetIndex])
            speechText += Message.ASK_NEXT
            handlerInput.responseBuilder
                    .withSpeech(speechText)
                    .withSimpleCard(Message.SKILL_NAME, speechText)
                    .withReprompt(Message.ASK_NEXT)
                    .withShouldEndSession(false)
                    .build()
        } else {
            // 最後の要素の場合
            getLastFund(handlerInput, sessionAttributes, found[targetIndex])
        }

    }

}
